### PR TITLE
ZCS-14611: User is able to delete account when license is in grace period

### PR DESCRIPTION
**Pull request by shubhamCS03**
Admin is able to delete the accounts when license is in grace period.
- Added the delete account check.

### CODE REVIEW

File: store/src/java/com/zimbra/cs/account/Provisioning.java [hunk 1]
```diff
@@ -1048,6 +1047,14 @@ public abstract Account restoreAccount(String emailAddress, String password,
      */
     public abstract void deleteAccount(String zimbraId) throws ServiceException;
 
+    /**
+     * Checking whether the license which is available in the server is valid or not
+     * @throws ServiceException
+     */
+    public void validateLicense() throws ServiceException{
```
Comment (ronstra-synacor) [inline store/src/java/com/zimbra/cs/account/Provisioning.java]:
```suggestion
    public void validateLicense() throws ServiceException {
```

File: store/src/java/com/zimbra/cs/account/Provisioning.java [hunk 2]
```diff
@@ -1048,6 +1047,14 @@ public abstract Account restoreAccount(String emailAddress, String password,
      */
     public abstract void deleteAccount(String zimbraId) throws ServiceException;
 
+    /**
+     * Checking whether the license which is available in the server is valid or not
+     * @throws ServiceException
```
Comment (ronstra-synacor) [inline store/src/java/com/zimbra/cs/account/Provisioning.java]:
```suggestion
     * @throws ServiceException if license is not valid, else return void
```

File: store/src/java/com/zimbra/cs/account/ldap/Validators.java [hunk 3]
```diff
@@ -67,7 +67,7 @@ private synchronized long getNextCheck() {
 
         @Override
         public void validate(Provisioning prov, String action, Object... args) throws ServiceException {
-            if (args.length < 1) return;
+            if (null == args || args.length < 1) return;
```
Comment (ronstra-synacor) [inline store/src/java/com/zimbra/cs/account/ldap/Validators.java]: I don't think its possible for varargs to be null. Did you run into an NPE?

**Review by ronstra-synacor — APPROVED** [review]

File: store/src/java/com/zimbra/cs/service/admin/DeleteAccount.java [hunk 4]
```diff
@@ -76,6 +74,7 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
         ZimbraSoapContext zsc = getZimbraSoapContext(context);
         Provisioning prov = Provisioning.getInstance();
 
+        prov.validateLicense();
```
Comment (RakeshAMore) [inline store/src/java/com/zimbra/cs/service/admin/DeleteAccount.java]: No.. I did not tell you to this..

File: store/src/java/com/zimbra/cs/account/Provisioning.java [hunk 5]
```diff
@@ -2628,6 +2635,7 @@ public static interface ProvisioningValidator {
         static final String RENAME_ACCOUNT_CHECK_DOMAIN_COS_AND_FEATURE =
             "renameAccountCheckDomainCosAndFeature";
         static final String DELETE_ACCOUNT_SUCCEEDED = "deleteAccountSucceeded";
+        static final String IS_VALID_LICENSE = "isValidLicense";
```
Comment (RakeshAMore) [inline store/src/java/com/zimbra/cs/account/Provisioning.java]: This is not what we discussed.

**Review by RakeshAMore — CHANGES_REQUESTED** [review]

File: store/src/java/com/zimbra/cs/account/Provisioning.java (hunk 1 above)
Comment (shubhamCS03) [reply store/src/java/com/zimbra/cs/account/Provisioning.java]: This code is removed now due to change in logic.

File: store/src/java/com/zimbra/cs/account/Provisioning.java (hunk 2 above)
Comment (shubhamCS03) [reply store/src/java/com/zimbra/cs/account/Provisioning.java]: This code has now been removed due to changes in logic.

File: store/src/java/com/zimbra/cs/account/ldap/Validators.java (hunk 3 above)
Comment (shubhamCS03) [reply store/src/java/com/zimbra/cs/account/ldap/Validators.java]: This code has now been removed due to changes in logic.

File: store/src/java/com/zimbra/cs/service/admin/DeleteAccount.java (hunk 4 above)
Comment (shubhamCS03) [reply store/src/java/com/zimbra/cs/service/admin/DeleteAccount.java]: Made changes as per our last discussion.

File: store/src/java/com/zimbra/cs/account/Provisioning.java (hunk 5 above)
Comment (shubhamCS03) [reply store/src/java/com/zimbra/cs/account/Provisioning.java]: Made changes as per our last discussion.

**Review by RakeshAMore — APPROVED** [review]

**Review by zimsuchitgupta — APPROVED** [review]

**Review by dasiyogesh — APPROVED** [review]

**Pull request merged**